### PR TITLE
Branch Merge

### PR DESCRIPTION
**Pull request by TypeVee**
Here is a compiled series of all previously reviewed branches, now with their comments resolved, formatted together for a merge into main

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,10 +1,40 @@
...
+        {res.status(404).send("Invalid Article")}
+      else if(err.constraint === "comments_author_fkey")
+        {res.status(403).send("Invalid Username")}
+      }
+    if(err.rowCount === 0)
+      {res.status(404).send('Article not found')}
+    else {res.status(500).send({msg: 'Internal Server Error'})}
+  });
 
+app.post('/api/articles/:article_id/comments')
```
Comment (TypeVee) [inline app.js:38]: Merge error, to be removed

File: __tests__/db.test.js [hunk 2]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+    test("Articles return sorted by their date in decending order", () =>{
+        return request(app)
+        .get('/api/articles')   
+        .then(({body})=>{
+            expect(new Date(body.articles[0].created_at).getTime() >= new Date(body.articles[1].created_at).getTime()).toBe(true)
+            expect(new Date(body.articles[1].created_at).getTime() >= new Date(body.articles[2].created_at).getTime()).toBe(true)
+            expect(new Date(body.articles[2].created_at).getTime() >= new Date(body.articles[3].created_at).getTime()).toBe(true)
+            expect(new Date(body.articles[3].created_at).getTime() >= new Date(body.articles[4].created_at).getTime()).toBe(true)
+            expect(new Date(body.articles[4].created_at).getTime() >= new Date(body.articles[6].created_at).getTime()).toBe(true)
+        })
```
Comment (NRMSMITH) [inline __tests__/db.test.js:158]: In order for this test to be accurate, you'd have to check every single article - which means about 13 expect statements. Using jest-sorted however, would mean that you would only have to use one expect statement, which is a much stronger test. Look at the documentation for jest-sorted to see how to implement it (I promise it's easier than this!)

File: __tests__/db.test.js [hunk 3]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+    })
+    test("Bad Queries do not effect outcome length or status", () =>{
+        return request(app)
+        .get('/api/articles?article_id=ga')
```
Comment (NRMSMITH) [inline __tests__/db.test.js:162]: The article_id wouldn't really be assigned this way - it generally would be assigned as the secondary part of the endpoint - i.e. '/api/articles/1'

File: __tests__/db.test.js [hunk 4]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+        .then((res)=>{
+                expect(res.statusCode).toBe(200)
+                expect(res.body.articles.length > 0)
+        })
+    })
+  })
+describe("GET /api/articles/", () => {
```
Comment (NRMSMITH) [inline __tests__/db.test.js:169]: make sure this describe is accurately describing what endpoint you are going to be testing

File: __tests__/db.test.js [hunk 5]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+            expect(res.statusCode).toBe(400)
+        })
+    })
+    test("Returns 404 status code when requesting an article ID with no data", () =>{
+        return request(app)
+        .get('/api/articles/135477')
+        .then((res)=>{
+            expect(res.statusCode).toBe(404)
+        })
+    })
```
Comment (NRMSMITH) [inline __tests__/db.test.js:197]: This article technically doesn't exist - so I would have a go at changing the describe

File: __tests__/db.test.js [hunk 6]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+        .then(({body})=>{
+            expect(body.article).toHaveProperty('author')
+            expect(body.article).toHaveProperty('title')
+            expect(body.article).toHaveProperty('article_id')
+            expect(body.article).toHaveProperty('body')
+            expect(body.article).toHaveProperty('topic')
+            expect(body.article).toHaveProperty('created_at')
+            expect(body.article).toHaveProperty('votes')
+            expect(body.article).toHaveProperty('article_img_url')
+        })
```
Comment (NRMSMITH) [inline __tests__/db.test.js:210]: The matcher .toMatchObject might be a bit more efficient here :)

File: __tests__/db.test.js [hunk 7]
```diff
@@ -52,8 +106,219 @@ describe("GET /api", () =>{
...
+                expect(res.statusCode).toBe(201)
+        })
+    })
+    test("Returns only the comment when successfully posted", ()=>{
+        return request(app).post("/api/articles/1/comments").send({username:"lurker", body:"Sorry -snip-"})
+            .then((res)=>{
+                expect(res.statusCode).toBe(201)
+                expect(res.body.postedComment).toBe("Sorry -snip-")
+                expect(Object.keys(res.body)).toEqual(["postedComment"])
+        })
```
Comment (NRMSMITH) [inline __tests__/db.test.js:227]: Realistically, you want to send back the entire comment, including username, body, comment_id, etc - everything that has now been successfully gathered from the db

**Review by NRMSMITH — CHANGES_REQUESTED** [review]
I've left quite a few comments on your testing suites - please make sure to implement the changes requested

File: __tests__/db.test.js (hunk 3 above)
Comment (TypeVee) [reply __tests__/db.test.js:162]: In this case, we're testing if queries could have any unintended functionality (Not params)
Main purpose was to see that the use of queries won't effect the outcome

**Pull request merged**